doc/SAIP4Extensions/SAI_P4EXT_API_Proposal.md

```diff
@@ -0,0 +1,279 @@
+
+} sai_p4ext_entry_attr_t;
+```
+
+
+### SAI P4 Extension APIs ###
+
+```cpp
+/**
+ * @brief Create an P4 table entry
+ *
+ * @param[out] p4ext_entry_id The P4 table id
+ * @param[in] switch_id The Switch Object id
+ * @param[in] attr_count Number of attributes
+ * @param[in] attr_list Array of attributes
+ *
+ * @return #SAI_STATUS_SUCCESS on success, failure status code on error
+ */
+typedef sai_status_t (*sai_create_p4ext_entry_fn)(
+        _Out_ sai_object_id_t *p4ext_entry_id,
+        _In_ sai_object_id_t switch_id,
+        _In_ uint32_t attr_count,
+        _In_ const sai_attribute_t *attr_list);
+
+/**
+ * @brief Delete an P4 entry
+ *
+ * @param[in] p4ext_entry_id The P4 table id
+ *
+ * @return #SAI_STATUS_SUCCESS on success, failure status code on error
+ */
+typedef sai_status_t (*sai_remove_p4ext_entry_fn)(
+        _In_ sai_object_id_t p4ext_entry_id);
+
+/**
+ * @brief Set P4 Table entry attribute
+ *
+ * @param[in] p4ext_entry_id The P4 table id
+ * @param[in] attr Attribute
+ *
+ * @return #SAI_STATUS_SUCCESS on success, failure status code on error
+ */
+typedef sai_status_t (*sai_set_p4ext_entry_attribute_fn)(
+        _In_ sai_object_id_t p4ext_entry_id,
+        _In_ const sai_attribute_t *attr);
+
+/**
+ * @brief Get P4 entry attribute
+ *
+ * @param[in] p4ext_entry_id P4 table id
+ * @param[in] attr_count Number of attributes
+ * @param[inout] attr_list Array of attributes
+ *
+ * @return #SAI_STATUS_SUCCESS on success, failure status code on error
+ */
+typedef sai_status_t (*sai_get_p4ext_entry_attribute_fn)(
+        _In_ sai_object_id_t p4ext_entry_id,
+        _In_ uint32_t attr_count,
+        _Inout_ sai_attribute_t *attr_list);
+```
+
+### SAI P4 Extension API Table ###
+
+```cpp
+/**
+ * @brief P4Ext  methods table retrieved with sai_api_query()
+ **/
+typedef struct _sai_p4ext_api_t
+{
+    sai_create_p4ext_entry_fn                     create_p4ext_entry;
+    sai_remove_p4ext_entry_fn                     remove_p4ext_entry;
+    sai_set_p4ext_entry_attribute_fn              set_p4ext_entry_attribute;
+    sai_get_p4ext_entry_attribute_fn              get_p4ext_entry_attribute;
+} sai_p4ext_api_t;
+
+```
+
+# API Usage #
+
+Consider a scenario where a Vendor A wishes to extend the SAI pipeline functionality with a custom feature not yet supported in SAI on a target that is P4 compliant. Let say this new feature classifies incoming packets based on packet field tuples SIP,DIP and sets the traffic class tc. As mentioned earlier the proposal does not deal with the implementation details of data pipeline by each individual vendor, thus the relative position of this table in the SAI pipeline is not part of the specification and is vendor dependent. Its upto the vendor to decide whether this table will come after the SAI QoS tables or before and resolve any conflict that would arise from the actions of the QoS tables and the new P4 Extension Table. This simple feature can be expressed in P4 as below.
```

Comment:
Why do we want to limit this extension to "P4 compliant" devices. I think this generic extension can be used to program features/tables in non-P4 devices as well.

Reply:
Alternate wording - "Consider a scenario where a Vendor A wishes to extend the SAI pipeline functionality with a custom feature which can be modeled with a P4 match-action table but is not yet supported in SAI."